### PR TITLE
add cookbook qda page

### PR DESCRIPTION
**Pull request by OXPHOS**
Ref #3105

### CODE REVIEW

**Comment by karlnapf** [comment]
I am not sure I see the connection between extracting things and the error here, can you explain in the chat?

**Comment by karlnapf** [comment]
travis worked

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 1]
```diff
@@ -0,0 +1,39 @@
+===============================
+Quadratic discriminant analysis
+===============================
+
+Quadratic discriminant analysis (QDA) is used in machine learning and statistical classification to separate measurements of two or more classes of objects or events by a quadric surface.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: remove "in machine learning and statistical classification"

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: remove "or events"

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+For QDA, the class label :math:`y` is assumed to be quadratic in the measurements of obeservations :math:`X`, so `y` will be decided based on:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: remove ", so y will be decided based on:" and replace be "i.e."

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+Example
+-------
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` as
+
+.. sgexample:: qda.sg:create_features
+
+
+We create an instance of the :sgclass:`CQDA` classifier with feature matrix and label list. 
+:sgclass:`CQDA` has two default arguments: `float:tolerance` for tolerance used in training and `bool:store_covs` marks whether to store the within class covariances
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: again, please dont explicitly mention variable names or methods in the text. Just say that there are parameters, and what they are for

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+.. sgexample:: qda.sg:create_instance
+
+Then we run the train QDA algorithm and apply it to test data, which here gives :sgclass:`CMulticlassLabels`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: Remove "then"

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+We can evaluate test performance via e.g. :sgclass:`CMulticlassAccuracy`.
+
+.. sgexample:: qda.sg:evaluate_accuracy
+
+
+----------
+References
+----------
+
+:wiki:`Quadratic_classifier`
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: Can you put the david barber citation for the QDA chapter her?
Also I think there is a wiki entry on QDA
You can also relate to LDA

File: doc/cookbook/source/examples/classifier/qda.rst (hunk 1 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: You can say that it is a generalisation of LDS

**Comment by karlnapf** [comment]
All good, same list as for other cookbook on hierarchical

File: doc/cookbook/source/examples/classifier/qda.rst (hunk 5 above)
Comment (OXPHOS) [inline doc/cookbook/source/examples/classifier/qda.rst]: Do you mean: Bayesian Reasoning and Machine Learning?

**Comment by OXPHOS** [comment]
@karlnapf I take back my words. So the problem actually is, when I try to get the cov matrix:
`RealMatrix c = qda.get_cov(classlabel)`

It gives error:
```
/Library/Ruby/Site/2.0.0/universal-darwin14/modshogun.bundle: warning: rb_define_const: invalid name `__MAC_10_8' for constant
[ERROR] In file /Users/zora/Github/shogun/src/shogun/io/File.cpp line 71: Error opening file '../../data/classifier_4class_2d_linear_features_train.dat'
ruby/classifier/qda.rb:3:in `initialize': [ERROR] In file /Users/zora/Github/shogun/src/shogun/io/File.cpp line 71: Error opening file '../../data/classifier_4class_2d_linear_features_train.dat' (fatal)
    from ruby/classifier/qda.rb:3:in `new'
    from ruby/classifier/qda.rb:3:in `<main>'
```

Which I cannot understand at all.
`qda.get_cov()` takes `int32_t` and doesn't require the access to origin dataset.

**Comment by OXPHOS** [comment]
Also the .sg cannot take `0`? Is this the way it should be or a bug?

When I tried to do 
```
#![extract_class_mean_and_cov]
int classlabel = 0
#![extract_class_mean_and_cov]

```
Syntax error in input: 
 on line 16
Traceback (most recent call last):
  File "/Users/zora/Github/shogun/examples/meta/generator/generate.py", line 116, in <module>
    includedTargets=args.targets, storeVars=storeVars)
  File "/Users/zora/Github/shogun/examples/meta/generator/generate.py", line 48, in translateExamples
    ast = parse(file.read(), os.path.join(dirRelative, filename))
  File "/Users/zora/Github/shogun/examples/meta/generator/parse.py", line 234, in parse
    return parser.parse(programString, filePath)
  File "/Users/zora/Github/shogun/examples/meta/generator/parse.py", line 32, in parse
    program = self.parser.parse(programString)
  File "/usr/local/lib/python2.7/site-packages/ply-3.8-py2.7.egg/ply/yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/local/lib/python2.7/site-packages/ply-3.8-py2.7.egg/ply/yacc.py", line 1049, in parseopt_notrack
    lookahead = get_token()     # Get the next token
  File "/usr/local/lib/python2.7/site-packages/ply-3.8-py2.7.egg/ply/lex.py", line 386, in token
    newtok = self.lexerrorf(tok)
  File "/Users/zora/Github/shogun/examples/meta/generator/parse.py", line 100, in t_error
    raise TypeError("Failed to tokenize input. Unknown text on line %d '%s'" % (t.lineno, t.value,))
TypeError: Failed to tokenize input. Unknown text on line 17 '0
#![extract_class_mean_and_cov]

#![evaluate_accuracy]
MulticlassAccuracy eval()
real accuracy = eval.evaluate(labels_predict, labels_test)
#![evaluate_accuracy]

# additional integration testing variables
RealVector output = labels_predict.get_labels()
'
make[3]: *** [examples/meta/CMakeFiles/meta_examples] Error 1
make[2]: *** [examples/meta/CMakeFiles/meta_examples.dir/all] Error 2
make[1]: *** [doc/cookbook/CMakeFiles/cookbook.dir/rule] Error 2
make: *** [cookbook] Error 2
```

File: doc/cookbook/source/examples/classifier/qda.rst (hunk 5 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: yes it is used in the other pages, e.g. knn.
There is a free online version, just reference the corresponding chapter like I did in KNN

**Comment by karlnapf** [comment]
Ok this needs investigation :)
Can you send me a gist with both problem illustrated, that I can put in my build and reproduce locally?
Thanks!

**Comment by OXPHOS** [comment]
@karlnapf 
get.cov() error (only in ruby): https://gist.github.com/OXPHOS/36a3c1f58cb657ab2f5ff86045238e6f
'0' error: https://gist.github.com/OXPHOS/4d58f4371e4e49ae8fbef16d176d8149

**Comment by karlnapf** [comment]
Ok got the error:
You constructed the QDA class with the (default) parameter `store_vars=false`
This means that you cannot extract covariances later.

I fixed that for you in #3220

**Comment by karlnapf** [comment]
This is a good example of "user(here you)  used shogun in a wrong way, but the error message is so cryptic that it doesnt help at all" ;)
We better make sure such a thing never happens ;)

**Comment by OXPHOS** [comment]
@karlnapf Thanks! Yes I didn't think about this at all, though I did know there's a default boolean parameter...

**Comment by OXPHOS** [comment]
@karlnapf Do you think we can put `tolerance` after `store_cov`, so users don't have to set `tolerance` to enable cov store? which one is more often used?

**Comment by karlnapf** [comment]
Feel free to send a separate patch with a new constructor for this case

**Comment by karlnapf** [comment]
Looks like you found another meta example bug ;)
I will fix it, which will make the Python build fail....

**Comment by karlnapf** [comment]
You have to use `True` with a capital T for the Python example to work

File: doc/cookbook/source/examples/classifier/qda.rst (hunk 5 above)
Comment (OXPHOS) [inline doc/cookbook/source/examples/classifier/qda.rst]: Sure.

File: doc/cookbook/source/examples/classifier/qda.rst [hunk 6]
```diff
@@ -0,0 +1,39 @@
+===============================
+Quadratic discriminant analysis
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/qda.rst]: btw can you rename all the files to their long name?
Sorry, but this is something that just came up in chat. Includes the data files (unfortunately)

**Comment by OXPHOS** [comment]
@karlnapf Only find a chapter that talks about LDA. Also, clicking the ref [Bar12]  links to `knn` page (So does Gaussian Naive Bayes page). Any way to solve it?

**Comment by karlnapf** [comment]
It is very closely related, that is why you can link to LDA as well.
Maybe also put this as a reference: https://en.wikipedia.org/wiki/Quadratic_classifier
You can also get inspiration in the description

**Comment by karlnapf** [comment]
Travis successfully executed the integration test https://travis-ci.org/shogun-toolbox/shogun/jobs/134037852#L3976

merging

**Pull request merged**